src/librustc/ty/sty.rs

```diff
@@ -1673,4 +1674,84 @@ pub struct Const<'tcx> {
     pub val: ConstVal<'tcx>,
 }
 
+impl<'tcx> Const<'tcx> {
+    pub fn from_bits(
+        tcx: TyCtxt<'_, 'tcx, '_>,
+        val: u128,
+        ty: Ty<'tcx>,
+    ) -> Self {
+        let alloc = Allocation::from_primval_ty(
+            tcx,
+            PrimVal::Bytes(val),
+            ty::ParamEnv::empty(),
+            ty);
+        Const {
+            val: ConstVal::Value(tcx.intern_const_alloc(alloc)),
+            ty,
+        }
+    }
+
+    pub fn zero_sized(tcx: TyCtxt<'_, 'tcx, '_>, ty: Ty<'tcx>) -> Self {
+        // FIXME: Pre-intern a zero-allocation
+        let alloc = Allocation::from_byte_aligned_bytes(&[]);
+        Const {
+            val: ConstVal::Value(tcx.intern_const_alloc(alloc)),
+            ty,
+        }
+    }
+
+    pub fn from_bool(tcx: TyCtxt<'_, 'tcx, '_>, v: bool) -> Self {
+        Const::from_bits(tcx, v as u128, tcx.types.bool)
+    }
+
+    pub fn from_usize(tcx: TyCtxt<'_, 'tcx, '_>, n: u64) -> Self {
+        Const::from_bits(tcx, n as u128, tcx.types.usize)
+    }
+
+    pub fn to_bits<'gcx>(&self, tcx: TyCtxt<'_, 'gcx, '_>, ty: Ty<'gcx>) -> Option<u128> {
+        if self.ty != ty {
```

Comment:
Can this be an assertion? In what cases to we actually return None due to a type mismatch?

Reply:
This just tries to convert a value to an u128 if it is of the correct type `ty`. It is used as an replacement for matching in many cases. An example is the `to_usize` method.